server/methods/core/cartToOrder.js

```diff
@@ -0,0 +1,212 @@
+ * @param {String} cartId - cartId to transform to order
+ * @return {String} returns orderId
+ */
+export function copyCartToOrder(cartId) {
+  check(cartId, String);
+  const cart = Collections.Cart.findOne(cartId);
+
+  // security check - method can only be called on own cart
+  if (cart.userId !== Meteor.userId()) {
+    throw new Meteor.Error(403, "Access Denied");
+  }
+
+  // Init new order object from existing cart
+  const order = Object.assign({}, cart);
+
+  // get sessionId from cart while cart is fresh
+  const sessionId = cart.sessionId;
+
+  // If there are no order items, throw an error. We won't create an empty order
+  if (!order.items || order.items.length === 0) {
+    const msg = "An error occurred saving the order. Missing cart items.";
+    Logger.error(msg);
+    throw new Meteor.Error("no-cart-items", msg);
+  }
+
+  // Debug only message to identify the current cartId
+  Logger.debug("cart/copyCartToOrder", cartId);
+
+  // Set our new order's cartId to existing cart._id
+  // We'll get a new _id for our order
+  order.cartId = cart._id;
+
+  // This block assigns an existing user's email associated with their account to this order
+  // We copied order from cart, so this userId and email are coming from the existing cart
+  if (order.userId && !order.email) {
+    // If we have a userId, but do _not_ have an email associated with this order
+    // we need to go find the account associated with this userId
+    const account = Collections.Accounts.findOne(order.userId);
+
+    // Check to make sure that the account exists and has an emails field
+    if (typeof account === "object" && account.emails) {
+      for (const email of account.emails) {
+        // If a user has specified an alternate "order" email address, use that
+        if (email.provides === "orders") {
+          order.email = email.address;
+          // Otherwise, check to see if the user has a "default" email address
+        } else if (email.provides === "default") {
+          order.email = email.address;
+        }
+        // If we can't find any relevant email addresses for the user, we'll
+        // let them assign an email address to this order after the checkout
+      }
+    }
+  }
+
+  // The schema will provide default values for these fields in our new order
+  // so we'll delete the values copied from the cart
+  delete order.createdAt; // autovalues from cart
+  delete order.updatedAt;
+  delete order.getCount;
+  delete order.getShippingTotal;
+  delete order.getSubTotal;
+  delete order.getTaxTotal;
+  delete order.getDiscounts;
+  delete order.getTotal;
+  delete order._id;
+
+
+  // Create a shipping record for each shop on the order
+  if (Array.isArray(order.shipping)) {
+    if (order.shipping.length > 0) {
+      const shippingRecords = [];
+      order.shipping.map((shippingRecord) => {
+        const billingRecord = order.billing.find(billing => billing.shopId === shippingRecord.shopId);
```

Comment:
This find only works with stripe unfortunately as other payment methods do not have a shopId.
Error below
```
Exception while invoking method 'cart/submitPayment' TypeError: Cannot read property '_id' of undefined
    at server/methods/core/cartToOrder.js:90:36
    at Array.map (native)
    at [object Object]._copyCartToOrder (server/methods/core/cartToOrder.js:88:22)
    at [object Object]._MethodHooks._wrappers.(anonymous function) (server/api/method-hooks.js:114:70)
    at packages/check.js:129:16
    at [object Object]._.extend.withValue (packages/meteor.js:1126:17)
    at Object.exports.Match._failIfArgumentsAreNotAllChecked (packages/check.js:128:41)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1765:18)
    at packages/ddp-server/livedata_server.js:1686:15
    at [object Object]._.extend.withValue (packages/meteor.js:1126:17)
    at packages/ddp-server/livedata_server.js:1684:36
    at [object Object]._.extend.applyAsync (packages/ddp-server/livedata_server.js:1683:12)
    at [object Object]._.extend.apply (packages/ddp-server/livedata_server.js:1622:26)
    at [object Object]._.extend.call (packages/ddp-server/livedata_server.js:1604:17)
    at [object Object].<anonymous> (server/methods/core/hooks/cart.js:31:32)
    at server/api/method-hooks.js:122:36
 => awaited here:
    at Promise.await (/Users/sn/.meteor/packages/promise/.0.8.9.1o4y3g5++os+web.browser+web.cordova/npm/node_modules/meteor-promise/promise_server.js:60:12)
    at [object Object]._.extend.apply (packages/ddp-server/livedata_server.js:1635:22)
    at [object Object]._.extend.call (packages/ddp-server/livedata_server.js:1604:17)
    at [object Object].<anonymous> (server/methods/core/hooks/cart.js:31:32)
    at server/api/method-hooks.js:122:36
    at arrayEach (/Users/sn/dev/reaction-dev/node_modules/lodash/_arrayEach.js:15:9)
    at forEach (/Users/sn/dev/reaction-dev/node_modules/lodash/forEach.js:38:10)
    at [object Object]._MethodHooks._wrappers.(anonymous function) (server/api/method-hooks.js:121:5)
    at packages/check.js:129:16
    at [object Object]._.extend.withValue (packages/meteor.js:1126:17)
    at Object.exports.Match._failIfArgumentsAreNotAllChecked (packages/check.js:128:41)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1765:18)
    at packages/ddp-server/livedata_server.js:719:19
    at [object Object]._.extend.withValue (packages/meteor.js:1126:17)
    at packages/ddp-server/livedata_server.js:717:46
    at [object Object]._.extend.withValue (packages/meteor.js:1126:17)
```